tests/codes/test_freecadapi.py

```diff
@@ -254,3 +255,138 @@ def func():
+        assert "Bluemira" in stp_content
+        assert (
+            "AP242_MANAGED_MODEL_BASED_3D_ENGINEERING_MIM_LF. {1 0 10303 442 1 1 4"
+            in stp_content
+        )
+
+        cadapi.save_cad(
+            [shape],
+            filename,
+            labels=["myshape"],
+            author="myfile",
+            stp_file_scheme="AP214IS",
+        )
+        assert os.path.exists(filename)
+        stp_content = open(filename).read()
+        assert "myshape" in stp_content  # shape label in file
+        assert "Bluemira" in stp_content  # bluemira still in file if author changed
+        assert "myfile" in stp_content  # author change
+        assert (
+            "AUTOMOTIVE_DESIGN { 1 0 10303 214 1 1 1 1 }" in stp_content
+        )  # scheme change
+
+
+class TestCADFiletype:
+    @classmethod
+    def setup_class(cls):
+        doc = newDocument()
+        cls.shape = doc.addObject("Part::Feature")
+        cls.shape.Shape = cadapi.extrude_shape(cadapi.make_circle(), (0, 0, 1))
+        doc.recompute()
+
+    def setup_method(self):
+        import FreeCADGui
+
+        if not hasattr(FreeCADGui, "subgraphFromObject"):
+            FreeCADGui.setupWithoutGUI()
+
+    @pytest.mark.parametrize("name, ftype", cadapi.CADFileType.__members__.items())
+    def test_init(self, name, ftype):
+        assert cadapi.CADFileType[name] == ftype
+        assert cadapi.CADFileType(ftype.value) == ftype
+
+    @pytest.mark.parametrize(
+        "name",
+        ("PDF", "VRML", "VRML_2", "VRML_ZIP", "VRML_ZIP_2", "WEBGL_X3D", "X3D", "X3DZ"),
+    )
+    def test_exporter_raised_FreeCADError(self, name, tmp_path):
+        """
+        These extensions require more of FreeCAD than we set up
+        Just tracking them.
+        Some of them may work if other packages are installed.
+        """
+
+        with pytest.raises(FreeCADError):
+            filetype = cadapi.CADFileType[name]
+            filetype.exporter([self.shape], f"{tmp_path/'tst'}.{filetype.value}")
+
+    @pytest.mark.parametrize(
+        "name",
+        (
+            "ACSII_STEREO_MESH",
+            "ADDITIVE_MANUFACTURING",
+            "AUTOCAD_DXF",
+            "BINMESH",
+            "BREP",
+            "BREP_2",
+            "CSG",
+            "FREECAD",
+            "GLTRANSMISSION",
+            "GLTRANSMISSION_2",
+            "IGES",
+            "IGES_2",
+            "INVENTOR_V2_1",
+            "JSON",
+            "JSON_MESH",
+            "OBJ",
+            "OBJ_WAVE",
+            "OFF",
+            "OPENSCAD",
+            "SIMPLE_MODEL",
+            "STEP",
+            "STEP_2",
+            "STL",
+            "THREED_MANUFACTURING",
+            # ifcopenshell package required
+            pytest.param("IFC_BIM", marks=[pytest.mark.xfail]),
+            pytest.param("IFC_BIM_JSON", marks=[pytest.mark.xfail]),
+            # Part.Feature has no compatible object type TODO find compatible object type
```

Comment:
Should we have these as options if we know they fail?

Reply:
My take was more if they do start working it would be nice to be aware. Maybe there is a better way of flagging this than loads of xfails?